### PR TITLE
Scheduled events endpoints

### LINKED ISSUE #30: Implement discords Guild Scheduled Event API
## As a user I want to...
be able to call the [Guild Scheduled Event](https://discord.com/developers/docs/resources/guild-scheduled-event) endpoints

## So that...
I can perform channel actions using my discord bot

## Context
Discord provides a rich set of APIs for interacting with guild scheduled events in a discord server. We should implement all these endpoints. This ticket covers only the Guild Scheduled Event endpoints. https://discord.com/developers/docs/resources/guild-scheduled-event

## Acceptance Criteria
GIVEN a user wants to send a request to interact with a Guild Scheduled Event on discord
THEN the specified action should be executed with sufficent error logging in place

### PR DESCRIPTION
**Pull request by nateweisz**
**Describe the PR**
Closes #30

### CODE REVIEW

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+public record CreateScheduledEventRequest(
+        String guildId,
+        Optional<String> channelId,
+        Optional<EntityMetadata> entityMetadata,
+        String name,
+        int privacyLevel, // 2 is only option atm
+        long scheduledStartTime, // ISO8601 timestamp (assuming long) note to self to check back
```
Comment (surajkumar) [inline src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java:17]: @Taz03 can you provide the correct format for `scheduledStartTime`

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.javadiscord.jdi.internal.api.DiscordRequest;
+import com.javadiscord.jdi.internal.api.DiscordRequestBuilder;
+import com.javadiscord.jdi.internal.models.guild.EntityMetadata;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/* NOTE TO FUTURE DEVS WHEN WE IMPLEMENT HIGHER LEVEL API
```
Comment (surajkumar) [inline src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java:11]: Let's create an issue to track this so we don't forget

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        Optional<String> image)
+        implements DiscordRequest {
+    @Override
+    public DiscordRequestBuilder create() {
+        Map<String, Object> body = new HashMap<>();
+
+        // check if entityType = 3 (EXTERNAL)
+        // channelId must be set to null
+        // entityMetadata (entity_metadata), scheduledEndTime are all required
+        if (entityType.isPresent() && entityType.get() == 3) {
```
Comment (surajkumar) [inline src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java:39]: Please can this if-check be moved into the constructor? Thanks

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            if (entityMetadata.isEmpty() || scheduledEndTime.isEmpty()) {
```
Comment (surajkumar) [inline src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java:40]: Please can this if-check be moved into the constructor? Thanks

**Review by surajkumar — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java (hunk 1 above)
Comment (Taz03) [reply src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java:17]: OffsetDateTime

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import com.javadiscord.jdi.internal.api.DiscordRequest;
+import com.javadiscord.jdi.internal.api.DiscordRequestBuilder;
+import com.javadiscord.jdi.internal.models.guild.EntityMetadata;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public record CreateScheduledEventRequest(
+        String guildId,
```
Comment (surajkumar) [inline src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java:12]: IDs are longs not strings!

**Review by surajkumar — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java (hunk 4 above)
Comment (nateweisz) [reply src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/ModifyScheduledEventRequest.java:33]: We discussed on discord and have decided to shelf this until the higher level api is implemented. That validation will stay for now.

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java (hunk 1 above)
Comment (nateweisz) [reply src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/CreateScheduledEventRequest.java]: Done!

**Review by surajkumar — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/GetScheduledEventUsersRequest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.javadiscord.jdi.internal.api.impl.scheduledevents;
+
+import com.javadiscord.jdi.internal.api.DiscordRequest;
+import com.javadiscord.jdi.internal.api.DiscordRequestBuilder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public record GetScheduledEventUsersRequest(
```
[the remark this answers is not in the document]
Comment (nateweisz) [reply src/main/java/com/javadiscord/jdi/internal/api/impl/scheduledevents/GetScheduledEventUsersRequest.java:10]: Fixed!

**Review by surajkumar — APPROVED** [review]

**Pull request merged**